nav2_dwb_controller/dwb_core/include/dwb_core/trajectory_critic.hpp

```diff
@@ -94,15 +95,24 @@ class TrajectoryCritic
   void initialize(
     const nav2_util::LifecycleNode::SharedPtr & nh,
     std::string & name,
-    std::shared_ptr<nav2_costmap_2d::Costmap2DROS> costmap_ros)
+    std::shared_ptr<nav2_costmap_2d::Costmap2DROS> costmap_ros,
+    std::shared_ptr<nav2_util::ParameterEventsSubscriber> param_sub = nullptr)
   {
     name_ = name;
     costmap_ros_ = costmap_ros;
     nh_ = nh;
+    param_subscriber_ = param_sub;
     if (!nh_->has_parameter(name_ + ".scale")) {
       nh_->declare_parameter(name_ + ".scale", rclcpp::ParameterValue(1.0));
     }
     nh_->get_parameter(name_ + ".scale", scale_);
+    if (param_sub) {
+      callback_handles_.push_back(param_subscriber_->add_parameter_callback(name_ + ".scale",
+        [&](const rclcpp::Parameter & p) {
+          std::lock_guard<std::recursive_mutex> lock(mutex_);
```

Comment:
you never take this mutex anywhere else in the class. I believe all the derived classes only use the lock in their parameter callbacks and they never sync with the main thread.

What are our options here?
1. Get rid of the lock and switch to atomics
2. Make the base class methods like `scoreTrajectory` `addGridScores` `getScale` and `setScale` take the lock. Rename them to something like `scoreTrajectory_sync` and have it call the derived class' `scoreTrajectory`.
3. Add a sync method to the interface where the controller explicitly syncs all variables prior to running an iteration of the loop.

Reply:
You're right. Any preference on options? I could also take the lock directly in the derived class implementations of those virtual functions.

Reply:
The atomics are relatively simple and clean. However, I don't know if it can be used in all cases, or if we can take the performance hit; if we use many atomics very frequently, the cost could be higher than a mutex.

I think my preference is to use the atomics where we can, and make the rest non-dynamic.